source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -0,0 +1,305 @@
+  }
+
+  // Write the remaining segment prepended by its length
+  count = name_.size() - last;
+  buffer_.writeBEInt<uint8_t>(count);
+  for (size_t i = 0; i < count; i++) {
+    buffer_.writeByte(*iter);
+    ++iter;
+  }
+
+  // Terminate the name record with a null byte
+  buffer_.writeByte(0x00);
+}
+
+// Serialize a DNS Query Record
+void DnsQueryRecord::serialize(Buffer::OwnedImpl& output) {
+  buffer_.drain(buffer_.length());
+
+  serializeName();
+  buffer_.writeBEInt<uint16_t>(type_);
+  buffer_.writeBEInt<uint16_t>(class_);
+
+  output.add(buffer_);
+}
+
+bool DnsMessageParser::parseDnsObject(const Buffer::InstancePtr& buffer) {
+
+  auto available_bytes = buffer->length();
+
+  memset(&incoming_, 0x00, sizeof(incoming_));
+
+  static constexpr uint64_t field_size = sizeof(uint16_t);
+  uint64_t offset = 0;
+  uint16_t data;
+
+  DnsQueryParseState state_{DnsQueryParseState::Init};
+
+  while (state_ != DnsQueryParseState::Finish) {
+
+    // Ensure that we have enough data remaining in the buffer to parse the query
+    if (available_bytes < field_size) {
+      ENVOY_LOG(error,
+                "Exhausted available bytes in the buffer. Insufficient data to parse query field.");
+      return false;
+    }
+
+    // Each aggregate DNS header field is 2 bytes wide.
+    data = buffer->peekBEInt<uint16_t>(offset);
+    offset += field_size;
+    available_bytes -= field_size;
+
+    if (offset > buffer->length()) {
+      ENVOY_LOG(error, "Buffer read offset [{}] is beyond buffer length [{}].", offset,
+                buffer->length());
+      return false;
+    }
+
+    switch (state_) {
+    case DnsQueryParseState::Init:
+      incoming_.id = data;
+      state_ = DnsQueryParseState::Flags;
+      break;
+
+    case DnsQueryParseState::Flags:
+      ::memcpy(static_cast<void*>(&incoming_.flags), &data, sizeof(uint16_t));
+      state_ = DnsQueryParseState::Questions;
+      break;
+
+    case DnsQueryParseState::Questions:
+      incoming_.questions = data;
+      state_ = DnsQueryParseState::Answers;
+      break;
+
+    case DnsQueryParseState::Answers:
+      incoming_.answers = data;
+      state_ = DnsQueryParseState::Authority;
+      break;
+
+    case DnsQueryParseState::Authority:
+      incoming_.authority_rrs = data;
+      state_ = DnsQueryParseState::Authority2;
+      break;
+
+    case DnsQueryParseState::Authority2:
+      incoming_.additional_rrs = data;
+      state_ = DnsQueryParseState::Finish;
+      break;
+
+    case DnsQueryParseState::Finish:
+      break;
+
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+
+  // Verify that we still have available data in the buffer to read answer and query records
+  if (offset > buffer->length()) {
+    ENVOY_LOG(error, "Buffer read offset[{}] is larget than buffer length [{}].", offset,
+              buffer->length());
+    return false;
+  }
+
+  // Each dns request has a Identification ID. This is used to match the request and replies.
```

Comment:
do we really care about the ID? 
there are 2 scenarios:
1. The filter already knows how to resolve this question - so in that case, there is no state to be maintained as we can immediately respond with the answer.
2. The filter issues a query to an external server - For this, the original query can be kept as a lambda state while invoking the dns_resolver API and response context is hence available.

IMO, this logic is unnecessary

Reply:
Yes.  I think it's necessary.   I envision a scenario where two queries are received by Envoy, one is resolved locally, the other must be resolved externally.   Since the query data is maintained in the filter, it would have knowledge of 2 queries for some period of time.  The ID is how it associates the query with the response.   True we can identify the response from the name in the query, but the purpose of the ID, per the RFC is to associate the query and response.

I prefer to keep this.

Reply:
Similar to my previous comment, I think maintaining class variables might be an overkill because there is no buffering to be handled. UDP provides all the data at once for a DNS request - so your request can be captured and responded back without using class variables. Would keep things much more cleaner and straightforward IMO.

Reply:
Also - when you actually get to implementing this async external query logic, you will see what I mean. This is just based on past experience of having done this.

Reply:
I can see what you mean.  I still believe it can be done with what I have.

I do have the async path already completed and working.   Let me see what I can do to migrate to what you are suggesting.

Reply:
Definitely. It can be done in multiple ways. I am just wondering if the code will be more easily comprehendible if there are less class members. 

Generally, the rule of thumb I follow when adding class methods or variables is to see if there is a way to achieve the desired outcome within reasonable limits by using static methods and non member variables. It allows for a more object oriented pattern as we are exposing less data to be "misused" in other code paths within the same class. I understand doing this may entail significant cost sometimes and wouldn't justify it, but that's the judgement call to me made imo.